shtab/__init__.py

```diff
@@ -558,18 +561,47 @@ def complete(
     """
     if isinstance(preamble, dict):
         preamble = preamble.get(shell, "")
-    if shell == "bash":
-        return complete_bash(
-            parser,
-            root_prefix=root_prefix,
-            preamble=preamble,
-            choice_functions=choice_functions,
-        )
-    if shell == "zsh":
-        return complete_zsh(
-            parser,
-            root_prefix=root_prefix,
-            preamble=preamble,
-            choice_functions=choice_functions,
+    try:
```

Comment:
As an alternative model and to enable the code to be extensible. It's useful to consider a registry model.

```python
__all__ = ['register_shell', 'get_supported_shells', 'get_completion_by_shell']

REGISTRY = {}

def register_shell(shell_id):
    def wrapper(func):
        if shell_id in REGISTRY:
            raise ValueError(f"{shell_id} is already registered")
        REGISTRY[shell_id] = func
        return func
    return wrapper


def get_supported_shells():
    return REGISTRY.keys()


def get_completion_by_shell(shell_id):
    return REGISTRY[shell_id]


@register_shell("zsh")
def complete_zsh(parser, root_prefix, preamble, choice_funcs):
    return "Zsh-Completion Shell str"


@register_shell("bash")
def complete_bash(parser, root_prefix, preamble, choice_funcs):
    return "Bash-Completion Shell str"
```

Reply:
Thanks for the registry idea.

FYI I haven't exposed `get_completion_by_shell` and `register_shell` because I can't think of a scenario where they'd be used.

AFAIK there are two scenarios where someone would write their own custom completer:

- `def complete_someprivshell()`: private org using a private/proprietary shell other than bash/zsh - wouldn't need `shtab`, they can just call `complete_someprivshell()` directly
- public FOSS completer, e.g. `def complete_powershell()`: should really just open a PR and add it to `shtab`

So in either case `shtab.register_shell` doesn't need to be exposed.